tsl/src/continuous_aggs/options.c

```diff
@@ -62,12 +67,166 @@ update_materialized_only(ContinuousAgg *agg, bool materialized_only)
 	ts_scan_iterator_close(&iterator);
 }
 
+/*
+ * Retrieve the cagg view query and find the groupby clause and
+ * time_bucket clause. Map them to the column names(of mat.hypertable)
+ * Returns: list of column names used in group by clause of the cagg query.
+ */
+static List *
+cagg_find_groupingcols(ContinuousAgg *agg, Hypertable *mat_ht)
+{
+	List *retlist = NIL;
+	ListCell *lc;
+	Oid cagg_view_oid =
+		get_relname_relid(NameStr(agg->data.user_view_name),
+						  get_namespace_oid(NameStr(agg->data.user_view_schema), false));
+	Relation cagg_view_rel = table_open(cagg_view_oid, AccessShareLock);
+	RuleLock *cagg_view_rules = cagg_view_rel->rd_rules;
+	Assert(cagg_view_rules && cagg_view_rules->numLocks == 1);
+	RewriteRule *rule = cagg_view_rules->rules[0];
+	if (rule->event != CMD_SELECT)
+		ereport(ERROR, (errcode(ERRCODE_TS_UNEXPECTED), errmsg("unexpected rule event for view")));
+
+	Query *cagg_view_query = copyObject(linitial(rule->actions));
+	table_close(cagg_view_rel, NoLock); // lock with be released at end of txn
+	Oid mat_relid = mat_ht->main_table_relid;
+	Query *orig_query;
+	/* the view rule has dummy old and new range table entries as the 1st and 2nd entries
+	 */
+	Assert(list_length(cagg_view_query->rtable) >= 2);
+	if (cagg_view_query->setOperations)
+	{
+		/* This corresponds to the union view.
+		 * the 3rd RTE entry has the SELECT 1 query from the union view. */
+		RangeTblEntry *orig_query_rte = lthird(cagg_view_query->rtable);
+		if (orig_query_rte->rtekind != RTE_SUBQUERY)
+			ereport(ERROR,
+					(errcode(ERRCODE_TS_UNEXPECTED),
+					 errmsg("unexpected rte type for view %d", orig_query_rte->rtekind)));
+
+		orig_query = orig_query_rte->subquery;
+	}
+	else
+	{
+		orig_query = cagg_view_query;
```

Comment:
`orig_query` is confusing me because it reminds me of the direct view. Maybe rephrase?

Reply:
renamed as finalize_query